bip-taro.mediawiki

```diff
@@ -392,18 +392,16 @@ derived using the first input outpoint as a sort of chain randomness to ensure
 uniqueness within the system as well as the output index and asset type to
 prevent collisions.
 The <code>asset_group_key</code> is derived as follows:
-* <code>asset_group_key</code> = <code>asset_key_internal +
-sha256(asset_key_internal || genesis_outpoint || output_index || asset_type)
+* <code>asset_group_key</code> = <code>TapTweak(asset_internal_group_key,
+ root)
 </code>
 
 where:
-* <code>asset_key_internal</code> is a 32-byte public key as defined by
+* <code>asset_internal_group_key = singleTweak(asset_raw_key, asset_id)</code> 
+* <code>asset_raw_key</code> is a 32-byte public key as defined by
 [[../master/bip-0340.mediawiki|BIP-340]]
-* <code>genesis_outpoint</code> is the first previous input outpoint used in the
-asset genesis transaction serialized in Bitcoin wire format.
-* <code>output_index</code> (4 byte, big-endian) is the index  of the output
-which contains the unique Taro commitment in the genesis transaction. 
-* <code>asset_type</code> (1 byte) is the type of the asset being minted.
+* <code>root<code> is a taproot that can hold arbitrary asset scripts enforcing
```

Comment:
`group_script_root`?

Reply:
One other thing I'd like us to consider here is the idea of also starting to timestamp the base/canonical universe information on chain. 

I've started to solidify the concept here: https://github.com/Roasbeef/bips/blob/bip-tap/bip-tap-universe.mediawiki#canonical-root-universe-state-transition-rules

At a high level, we'd:
  * Add a new field to opt into enforced chain commitments of the Universe root (total supply)
  * If set, then the first output in the minting transaction must use the asset group key (in the context of this diff, I think it'd be the `asset_raw_key`) as the internal key of that output. 
  * That output would then commit to the new universe root (modified commitment) that includes the issued asset

Then for each other asset:
  * You can only mint by spending that output created above (linked issuance)
  * The first output inherits the above, and also should be a new valid commitment 

So basically trying to re-use the group key, such that you also need to satisfy it in order to do further issuance. Then this way, ppl can watch a single output on chain to know when more assets are issued (run the numbers, etc, etc). 

The goal here is that we'd sort of _force them_ (ideally we also support update/rotation of the policy too) to use the exact same tapscript root as the tap layer on the Bitcoin layer. This means you need to be able to move the Bitcoin output to make a new tap asset in the group and the other way around. 

I think we can achieve this if we add some verification/restrictions on the tapscript sibling hash (so the thing to the left/right of the tap/universe commitment). 

Other variations exists too: 
  * We can derive a NUMS key from the OG group key, then use that to always force script reveal (that sibling hash is revealed as you go to the left/right of the tree where all the scripts live). 

I think it's still nice to also retain this new tap witness requirement on the asset layer as you can have other functionality not yet deployed on script (other sigs, zkps, etc, etc).

Reply:
> Doesn't this only force use of the same internal key, but not the actual tapscript root & tree? So you could use different trees on the Bitcoin and Tap layers (not sure why you'd want to though).



Correct that it's just the internal key that's being enforced. Agree you'd also want to have it be mirrored as much as possible, which would then also require a proof of tapscript sub-tree equivalence.